Add binder resolution events

lpereira opened the pull request:
This addresses comments made in the coreclr repo; for more information, please refer to [issue 27862](https://github.com/dotnet/coreclr/pull/27862).

lpereira committed:
Add assembly resolution attempt events

These events are fired every time there's an attempt at resolving an
assembly before it is loaded.  Events are fired when trying to find the
assembly in the current load context, using a custom
AssemblyLoadContext, looking in the TPA, and so on.  Results are either
success, or a more descriptive value (e.g. assembly not found, or
incompatible version).

lpereira committed:
Maintain thread-local cache of AssemblyLoadContext name

vitek-karas commented on src/coreclr/src/binder/bindertracing.cpp:
```diff
@@ -231,3 +259,105 @@ namespace BinderTracing
+            }
+        }
+
+        StackSString assemblyLoadContext;
```
Nit: please rename to assemblyLoadContextName

vitek-karas commented on src/coreclr/src/binder/clrprivbinderassemblyloadcontext.cpp:
```diff
@@ -57,24 +60,35 @@ HRESULT CLRPrivBinderAssemblyLoadContext::BindAssemblyByName(IAssemblyName     *
+
         // Step 1 - Try to find the assembly within the LoadContext.
+
+        tracer.SetAssemblyName(pAssemblyName);
```
This is repeating line 63 - probably unnecesarilly.

vitek-karas commented on src/coreclr/src/binder/clrprivbinderassemblyloadcontext.cpp:
```diff
@@ -57,24 +60,35 @@ HRESULT CLRPrivBinderAssemblyLoadContext::BindAssemblyByName(IAssemblyName     *
         {
+            INT_PTR managedAssemblyLoadContext = GetManagedAssemblyLoadContext();
+
+            tracer.SetManagedALC(managedAssemblyLoadContext);
```
Does this mean that the FindInLoadContext stage would not indicate which ALC it was using?

vitek-karas commented on src/coreclr/src/binder/clrprivbinderassemblyloadcontext.cpp:
```diff
@@ -57,24 +60,35 @@ HRESULT CLRPrivBinderAssemblyLoadContext::BindAssemblyByName(IAssemblyName     *
+            INT_PTR managedAssemblyLoadContext = GetManagedAssemblyLoadContext();
+
+            tracer.SetManagedALC(managedAssemblyLoadContext);
+            tracer.GoToStage(BinderTracing::ResolutionAttemptedOperation::Stage::FindInLoadContext);
```
I still don't see the logic of calling GoToStage with the exact same stage as before... Does this mean we fire the same event twice... ?

vitek-karas commented on src/coreclr/src/vm/appdomain.cpp:
```diff
@@ -6243,6 +6247,8 @@ HRESULT RuntimeInvokeHostAssemblyResolver(INT_PTR pManagedAssemblyLoadContextToB
     {
         bool fResolvedAssembly = false;
 
+        tracer.SetManagedALC(pManagedAssemblyLoadContextToBindWithin);
```
Why not set this right where we set the name - the ALC is available there as well.

vitek-karas commented on src/coreclr/src/vm/appdomain.cpp:
```diff
@@ -6255,6 +6261,7 @@ HRESULT RuntimeInvokeHostAssemblyResolver(INT_PTR pManagedAssemblyLoadContextToB
         {
             // Step 2 (of CLRPrivBinderAssemblyLoadContext::BindUsingAssemblyName) - Invoke Load method
             // This is not invoked for TPA Binder since it always returns NULL.
+            tracer.GoToStage(BinderTracing::ResolutionAttemptedOperation::Stage::FindInLoadContext);
```
I imagine this should be AssemblyLoadContextLoad stage - right?

vitek-karas commented on src/coreclr/src/binder/bindertracing.cpp:
```diff
@@ -80,16 +80,48 @@ namespace
+        // Keeping a per-thread INT_PTR reference to a managed AssemblyLoadContext is
+        // sufficient to avoid repeatedly calling into managed code, as managedALC does not
+        // change between each resolve+bind operations.
+        static thread_local INT_PTR prevManagedALC;
```
To be honest I'm not a fan of this solution. While it will probably work it feels too "fragile". There are basically two callers for this functionality:
* `BinderTracing PopulateBindRequest` which caches the results on its `BindRequest` object.
* The newly added `ResolutionAttemptedOperation` which can cache the results since it's also instance based.

Caching the results feel like a more natural solution to the problem. Not counting that with the change as is, one caller is caching and the other is not.

vitek-karas commented on src/coreclr/src/binder/bindertracing.cpp:
```diff
@@ -231,3 +259,105 @@ namespace BinderTracing
+        }
+
+        StackSString errorMsg, resultAssemblyName, resultAssemblyPath;
+        uint16_t result;
```
Nit: Declare this as the enum type so that instead of casting on each assignment, you can cast only once at the end when using it.

vitek-karas reviewed (commented):
* I would very much like to see a sample of the traces for a given bind operation (the typical success case, and some interesting fallback and failure case).
* While I understand that having a Start for each stage is nice from profiling point of view, for diagnostics it doesn't really help with anything as far as I can tell. I think we should log just at the end. The flow chart of how stages go one after the other is something we need to document anyway (it's basically the binder algorithm description), so it's perfectly OK to expect users of tracing to have a rough 
understanding of the various stages. I fear that if we fire too many events it will make it harder to use (especially without special tooling - basically a textual dump of events will be unnecessarily messy, and special tooling would combine the Start/End into one line anyway for readability).
* Missing tests

elinor-fung reviewed (commented):
I agree with @vitek-karas about just logging the end. I think we should avoid the extra events right now. And if we do find a compelling reason to add them later, it should be simple.

Adding tests should also help clarify/catch some of the questions around potential missing properties or duplicate traces.

elinor-fung committed:
Stop firing event for start of stage

Fix ClrEtwAll.man

elinor-fung committed:
Cache assembly and ALC name on ResolutionAttemptedOperation instance

Remove duplicate stage tracing
Trace stages in RuntimeInvokeHostAssemblyResolver

TODO: FindInLoadContext, PlatformAssemblies, AppDomainAssemblyResolveEvent stages

elinor-fung committed:
Trace stages (FindInLoadContext, PlatformAssemblies) in AssemblyBinder

TODO: AppDomainAssemblyResolveEvent, exceptions

elinor-fung committed:
Add tests

elinor-fung committed:
Add resolution result for exception thrown during stage

Add tests for exception thrown in extension points

elinor-fung committed:
Trace AppDomainAssemblyResolveEvent stage

Add tests

elinor-fung commented:
This should now have all the stages traced out. A bunch has changed, so I resolved all the comments that were referring to parts that are no longer there - please take a look.

General flow (where if one stage succeeds, the ones below it would not happen):

Default ALC
```
FindInLoadContext                    (DefaultALC)
PlatformAssemblies                   (DefaultALC)
AssemblyLoadContextResolvingEvent    (DefaultALC)
AppDomainAssemblyResolveEvent        (DefaultALC)
```
Custom ALC
```
FindInLoadContext                    (CustomALC)
AssemblyLoadContextLoad              (CustomALC)
FindInLoadContext                    (DefaultALC)
PlatformAssemblies                   (DefaultALC)
AssemblyLoadContextResolvingEvent    (DefaultALC)*
DefaultAssemblyLoadContextFallback   (CustomALC)
AssemblyLoadContextResolvingEvent    (CustomALC)
AppDomainAssemblyResolveEvent        (CustomALC)

* The existence of this stage is dependent on whether or not AssemblyLoadContext.Default has already been initialized (related: corefx#42477)
```

For incompatible version or name mismatch failures in `FindInLoadContext` or `PlatformAssemblies`, the remaining resolution stages are still performed.

Known issues that I would like to handle in a follow-up change:
- error cases:
  - for Default ALC, incompatible version in FindInLoadContext or PlatformAssemblies – gets reported as name mismatch
  - loading by assembly path where assembly with same name but different MVID is already loaded – gets reported as a success in FindInLoadContext, then a failure for the entire bind operation (the exception thrown was already special-cased to have more info)
- tests for `ResolveSatelliteAssembly` stage

Some things I realized that were not obvious to me:
- Satellite assembly
  - loaded from culture subdirectory of parent assembly – doesn’t actually validate the culture of the loaded assembly
  - platform resource root and app path search is always for the exact culture casing
- Extension points
  - no validation of version / culture of the returned assembly
  - exception in ALC.Resolving event handler doesn’t set inner exception if it is for the default ALC. It is set for custom ALC.
  - if ALC.Load or ALC.Resolving throw FileNotFoundException, AppDomain.AssemblyResolve is still fired. Any other exception and it is not fired.

vitek-karas commented:
> Some things I realized that were not obvious to me:
Can you please create issues for each of these? They are very interesting and we should follow up on each. Some of them are likely fixable, others are potential breaking changes, so maybe we won't do anything. But in any case we should have issues so that we can backfill tests and docs about these behaviors.

vitek-karas commented on src/coreclr/src/binder/bindertracing.cpp:
```diff
@@ -231,3 +271,178 @@ namespace BinderTracing
+
+namespace BinderTracing
+{
+    ResolutionAttemptedOperation::ResolutionAttemptedOperation(AssemblyName *assemblyName, UINT_PTR binderID, const HRESULT& hr)
```
These two overloads feel a bit dangerous - INT_PTR versus UINT_PTR - the two types feel a bit too similar. I would not mind if we had just the one which takes both binderID and managedALC - it makes things explicit.

vitek-karas commented on src/coreclr/src/binder/bindertracing.cpp:
```diff
@@ -231,3 +271,178 @@ namespace BinderTracing
+        if (!BinderTracing::IsEnabled() || stage == Stage::NotYetStarted)
+            return;
+
+        if (!m_populatedAttemptInfo)
```
This seems to always be true since we set it in the ctor...

elinor-fung replied:
This was to handle if tracing was enabled after the class was constructed but before all the stages were done. It would also be reasonable (and make things simpler - this was the reason for the existence of the AttemptInfo struct as well) to say that if tracing wasn't enabled at construction, the class doesn't trace anything.

vitek-karas replied:
Binding operation is naturally a short-lived thing - and typically pretty fast. So for "monitoring" scenarios I don't think it would matter at all. It might be a tiny little bit interesting for live-debugging scenarios, but I don't really see that as a thing. If it simplifies the code - let's not support "in the middle enable"...

elinor-fung replied:
Only tracing if it was enabled at construction of this class now.

vitek-karas commented on src/coreclr/src/binder/inc/bindertracing.h:
```diff
@@ -46,9 +53,128 @@ namespace BinderTracing
+            NotYetStarted = 0xffff, // Used as flag to not fire event; not present in value map
+        };
+
+        struct AttemptInfo
```
I don't see a need for this structure to be honest. We don't seem to expose it outside of the operation class and it's only used as a field which is always populated in the ctor. I would simply add the necessary fields of it onto the parent class.

vitek-karas commented on src/coreclr/src/binder/bindertracing.cpp:
```diff
@@ -231,3 +271,178 @@ namespace BinderTracing
+    // static
+    void ResolutionAttemptedOperation::TraceAppDomainAssemblyResolve(AssemblySpec *spec, PEAssembly *resultAssembly, Exception *exception)
+    {
+        Result result;
```
This should not do any work if the tracing is not enabled.

elinor-fung replied:
🤦‍♀

vitek-karas commented on src/coreclr/src/binder/inc/bindertracing.h:
```diff
@@ -46,9 +53,128 @@ namespace BinderTracing
+
+        void SetException(Exception *ex)
+        {
+            ex->GetMessage(m_exceptionMessage);
```
I would think this should not do any work if tracing is not enabled (this makes a string copy - so not necessarily a cheap operation).

vitek-karas commented on src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.ResolutionFlow.cs:
```diff
@@ -0,0 +1,586 @@
+            }
+        }
+
+        // Successful load through AssemblyLoadContext.Resolving event (Custom ALC):
```
Exception in ...

vitek-karas commented on src/coreclr/tests/src/Loader/binding/tracing/BinderTracingTest.ResolutionFlow.cs:
```diff
@@ -0,0 +1,586 @@
+            }
+        }
+
+        // Successful load through AssemblyLoadContext.Resolving event (default ALC):
```
AppDomain..ResolveEvent

elinor-fung commented on src/coreclr/src/binder/assemblybinder.cpp:
```diff
@@ -303,6 +304,11 @@ namespace BINDER_SPACE
+        // Tracing happens outside the binder lock to avoid calling into managed code within the lock
+        NewHolder<BinderTracing::ResolutionAttemptedOperation> tracer;
+        if (pAssemblyName != nullptr)
+            tracer = new BinderTracing::ResolutionAttemptedOperation(pAssemblyName, pApplicationContext->GetBinderID(), hr);
```
Yeah, I wasn't immediately sure how I would want this to behave (and I didn't find usage where tracing seemed very valuable, so I put off dealing with it). An empty AssemblyName (or using the file name/path instead) seemed a bit weird, and adding a property for the path for just one 'real' use case seemed undesirable as well. I think using `szCodeBase` for the name would be the most reasonable option.

Aside from the main assembly, the only thing I saw that I think would go through this is crossgen? (didn't actually debug through it - just looked at callers)

vitek-karas replied:
I have actually two concerns here:
* Consistency - if we only trace sometimes, it makes things inherently more complex (and harder to diagnose)
* Simplicity/perf - the code now adds a heap allocation for basically all cases except a very few rare... so we're paying a higher price by everyone for a tiny minority to take advantage of

Given that this really only kicks in for main assembly - I would be perfectly OK if we simply traced empty name - I know it's not the best - but we can easily tell which one it is.

elinor-fung replied:
Just tracing empty name now.

elinor-fung committed:
Only trace if enabled at construction of ResolutionAttemptedOperation

elinor-fung committed:
PR feedback

elinor-fung committed:
Rename PlatformAssemblies stage -> ApplicationAssemblies

elinor-fung commented:
Filed:
- Satellite assembly search in platform resource roots only uses exact culture casing #1019
- Culture of assembly is not validated when loaded from subdirectory of parent assembly #1021
- Version and culture of assembly returned by assembly loading extension points are not validated #1022
- Inner exception is not set when AssemblyLoadContext.Resolving handler throws an exception #1023
- AppDomain.AssemblyResolve can be fired/not fired depending on the type of exception thrown by other extension points #1025

vitek-karas reviewed (approved):
LGTM
